fem/transfer.cpp

```diff
@@ -871,13 +868,204 @@ void L2ProjectionGridTransfer::L2ProjectionH1Space::AllocR()
+{
+   delete R_sm;
+   delete M_LH_sm;
+   delete RTxM_LH_sm;
+}
+
+void L2ProjectionGridTransfer::SerialL2ProjectionH1Space::LumpedMassInverse(
+   Vector& ML_inv) const
+{
+   for (int i = 0; i < ML_inv.Size(); ++i)
+   {
+      ML_inv[i] = 1.0 / ML_inv[i];
+   }
+}
+
+void L2ProjectionGridTransfer::SerialL2ProjectionH1Space::GetHOTDofsByVDim(
+   int vdim, const Vector& x, Vector& x_vdim_true) const
+{
+   GetDofsByVDim(fes_ho, vdim, x, x_vdim_true);
+}
+
+void L2ProjectionGridTransfer::SerialL2ProjectionH1Space::GetLORTDofsByVDim(
+   int vdim, const Vector& x, Vector& x_vdim_true) const
+{
+   GetDofsByVDim(fes_lor, vdim, x, x_vdim_true);
+}
+
+void L2ProjectionGridTransfer::SerialL2ProjectionH1Space::SetHOFromTDofsByVDim(
+   int vdim, const Vector& y_vdim_true, Vector& y) const
+{
+   SetFromDofsByVDim(fes_ho, vdim, y_vdim_true, y);
+}
+
+void L2ProjectionGridTransfer::SerialL2ProjectionH1Space::SetLORFromTDofsByVDim(
+   int vdim, const Vector& y_vdim_true, Vector& y) const
+{
+   SetFromDofsByVDim(fes_lor, vdim, y_vdim_true, y);
+}
+
+#ifdef MFEM_USE_MPI
+
+L2ProjectionGridTransfer::ParL2ProjectionH1Space::ParL2ProjectionH1Space(
+   const ParFiniteElementSpace& pfes_ho_, const ParFiniteElementSpace& pfes_lor_)
+   : L2ProjectionH1Space(pfes_ho_, pfes_lor_),
+     pfes_ho(pfes_ho_),
+     pfes_lor(pfes_lor_)
+{
+   if (pfes_ho.GetVDim() != 1)
+   {
+      pfes_ho_scalar = new ParFiniteElementSpace(pfes_ho.GetParMesh(),
+                                                 pfes_ho.FEColl(), 1);
+      pfes_lor_scalar = new ParFiniteElementSpace(pfes_lor.GetParMesh(),
+                                                  pfes_lor.FEColl(), 1);
+   }
+   else // vdim == 1
+   {
+      pfes_ho_scalar = &pfes_ho;
+      pfes_lor_scalar = &pfes_lor;
+   }
+   SparseMatrix* R_sm;
+   SparseMatrix* M_LH_sm;
+   std::tie(R_sm, M_LH_sm) = ComputeSparseRAndM_LH();
+
+   HypreParMatrix R_local = HypreParMatrix(pfes_ho_scalar->GetComm(),
+                                           pfes_lor_scalar->GlobalVSize(), pfes_ho_scalar->GlobalVSize(),
+                                           pfes_lor_scalar->GetDofOffsets(), pfes_ho_scalar->GetDofOffsets(), R_sm);
+   HypreParMatrix M_LH_local = HypreParMatrix(pfes_ho_scalar->GetComm(),
+                                              pfes_lor_scalar->GlobalVSize(), pfes_ho_scalar->GlobalVSize(),
+                                              pfes_lor_scalar->GetDofOffsets(), pfes_ho_scalar->GetDofOffsets(), M_LH_sm);
+
+   R_par = RAP(pfes_lor_scalar->Dof_TrueDof_Matrix(), &R_local,
+               pfes_ho_scalar->Dof_TrueDof_Matrix());
+   M_LH_par = RAP(pfes_lor_scalar->Dof_TrueDof_Matrix(), &M_LH_local,
+                  pfes_ho_scalar->Dof_TrueDof_Matrix());
+   HypreParMatrix* R_T = R_par->Transpose();
+   RTxM_LH_par = ParMult(R_T, M_LH_par, true);
+
+   R = R_par;
+   M_LH = M_LH_par;
+   RTxM_LH = RTxM_LH_par;
+
+   pcg = CGSolver(MPI_COMM_WORLD);
+   InitializeCGSolver();
```

Comment:
Interesting, I guess this calls the implicitly defined move constructor of `CGSolver`. Seems like that should work…

I would prefer having a the constructor of `L2ProjectionH1Space` initialize the `CGSolver` (and set the default options, thereby doing away with `InitializeCGSolver` altogether). I guess this means there will have to be an `#ifdef`'d constructor that takes the `MPI_Comm` (just like there is for `CGSolver` itself).